lib/obj.js

```diff
@@ -1,365 +1,479 @@
-            } else {
-                currentIndexArray = materialGroups[material];
-                if (!defined(currentIndexArray)) {
-                    currentIndexArray = [];
-                    materialGroups[material] = currentIndexArray;
-                }
-            }
+function Mesh() {
+    this.name = undefined;
+    this.primitives = [];
+    this.positions = new ArrayStorage(ComponentDatatype.FLOAT);
+    this.normals = new ArrayStorage(ComponentDatatype.FLOAT);
+    this.uvs = new ArrayStorage(ComponentDatatype.FLOAT);
+}
+
+function Primitive() {
+    this.material = undefined;
+    this.indices = new ArrayStorage(ComponentDatatype.UNSIGNED_INT);
+}
+
+// OBJ regex patterns are modified from ThreeJS (https://github.com/mrdoob/three.js/blob/master/examples/js/loaders/OBJLoader.js)
+var vertexPattern = /v( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)/;                                                      // v float float float
+var normalPattern = /vn( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)/;                                                     // vn float float float
+var uvPattern = /vt( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)/;                                                                               // vt float float
+var facePattern1 = /f( +-?\d+)\/?( +-?\d+)\/?( +-?\d+)\/?( +-?\d+)?\/?/;                                                                        // f vertex vertex vertex ...
+var facePattern2 = /f( +(-?\d+)\/(-?\d+)\/?)( +(-?\d+)\/(-?\d+)\/?)( +(-?\d+)\/(-?\d+)\/?)( +(-?\d+)\/(-?\d+)\/?)?/;                            // f vertex/uv vertex/uv vertex/uv ...
+var facePattern3 = /f( +(-?\d+)\/(-?\d+)\/(-?\d+))( +(-?\d+)\/(-?\d+)\/(-?\d+))( +(-?\d+)\/(-?\d+)\/(-?\d+))( +(-?\d+)\/(-?\d+)\/(-?\d+))?/;    // f vertex/uv/normal vertex/uv/normal vertex/uv/normal ...
+var facePattern4 = /f( +(-?\d+)\/\/(-?\d+))( +(-?\d+)\/\/(-?\d+))( +(-?\d+)\/\/(-?\d+))( +(-?\d+)\/\/(-?\d+))?/;                                // f vertex//normal vertex//normal vertex//normal ...
+
+/**
+ * Parse an obj file.
+ *
+ * @param {String} objPath Path to the obj file.
+ * @param {Object} [options] An object with the following properties:
+ * @param {Boolean} [options.hasTransparency=false] Do a more exhaustive check for texture transparency by looking at the alpha channel of each pixel.
+ * @param {Boolean} [options.secure=false] Prevent the converter from reading image or mtl files outside of the input obj directory.
+ * @param {Boolean} [options.logger] A callback function for handling logged messages. Defaults to console.log.
+ * @returns {Promise} A promise resolving to the obj data.
+ * @exception {RuntimeError} The file does not have any geometry information in it.
+ *
+ * @private
+ */
+function loadObj(objPath, options) {
+    // The defaults are set in convert as well, this just helps with testing loadObj individually
+    options = combine(options, {
```

Comment:
No need for `Cesium.combine` here, use `Object.assign`
```javascript
options = Object.assign({
    hasTransparency : false,
    secure : false,
    logger : function(message) {
        console.log(message);
    }
}, options);
```

Also, extract the default logger function out into a module scope variable, no need to create a new instance every time.